pulsar-broker/src/main/java/org/apache/pulsar/broker/service/persistent/PersistentSubscription.java

```diff
@@ -971,19 +971,24 @@ public synchronized void redeliverUnacknowledgedMessages(Consumer consumer, List
                     (PositionImpl) this.pendingCumulativeAckMessage;
 
             positions.forEach(position -> {
-                if ((pendingAckMessages == null || (pendingAckMessages != null &&
-                        !this.pendingAckMessages.contains(position))) &&
-                        (null == cumulativeAckPosition ||
-                                (null != cumulativeAckPosition && position.compareTo(cumulativeAckPosition) > 0))) {
+                if ((pendingAckMessages == null || !this.pendingAckMessages.contains(position))
+                        && (null == cumulativeAckPosition || position.compareTo(cumulativeAckPosition) > 0)) {
                     pendingPositions.add(position);
                 }
             });
-            dispatcher.redeliverUnacknowledgedMessages(consumer, pendingPositions);
+            dispatcher.redeliverUnacknowledgedMessages(consumer, trimByMarkDeletePosition(pendingPositions));
         } else {
-            dispatcher.redeliverUnacknowledgedMessages(consumer, positions);
+            dispatcher.redeliverUnacknowledgedMessages(consumer, trimByMarkDeletePosition(positions));
         }
     }
 
+    private List<PositionImpl> trimByMarkDeletePosition(List<PositionImpl> positions) {
```

Comment:
This is applying this filtering which generates a lot of temp objects each time these redeliveries are done. Why not clear the map when the clear-backlog or seek commands are issued?

Reply:
Replace with removeIf(), please take a look.